File: examples/DOM/dom_interactions.rb
# UI demo: create a styled div with gradient + shadow, and wire interactions

# Container full-screen with gradient background
container = document.createElement('div')
container['style']['width'] = '100%'
container['style']['height'] = '100vh'
container['style']['margin'] = '0'
container['style']['display'] = 'flex'
container['style']['alignItems'] = 'center'
container['style']['justifyContent'] = 'center'
container['style']['background'] = 'linear-gradient(135deg, #0ea5e9 0%, #6366f1 100%)'
container['style']['position'] = 'relative'
document['body'].appendChild(container)

# Draggable, resizable card
box = document.createElement('div')
box['style']['width'] = '220px'
box['style']['height'] = '140px'
box['style']['borderRadius'] = '16px'
box['style']['boxShadow'] = '0 10px 25px rgba(0,0,0,.25)'
box['style']['background'] = 'linear-gradient(135deg, #6EE7F9 0%, #9333EA 100%)'
box['style']['position'] = 'relative'
box['style']['cursor'] = 'grab'
box['style']['userSelect'] = 'none'
box['style']['touchAction'] = 'none'
box['style']['willChange'] = 'transform'
box['style']['transform'] = 'translate3d(0px, 0px, 0)'
box['style']['transition'] = 'transform .2s ease, width .2s ease, height .2s ease, box-shadow .2s ease'
default_transition = box['style']['transition']
container.appendChild(box)

label = document.createElement('div')
label['style']['color'] = '#fff'
label['style']['fontFamily'] = 'system-ui, -apple-system, Segoe UI, Roboto, sans-serif'
label['style']['fontSize'] = '14px'
label['style']['padding'] = '12px'
label['innerText'] = 'Click / Tap / Drag me'
box.appendChild(label)

# Resize handle (bottom-right)
handle = document.createElement('div')
handle['style']['position'] = 'absolute'
handle['style']['right'] = '6px'
handle['style']['bottom'] = '6px'
handle['style']['width'] = '14px'
handle['style']['height'] = '14px'
handle['style']['borderRadius'] = '3px'
handle['style']['background'] = 'rgba(255,255,255,.85)'
handle['style']['boxShadow'] = '0 1px 3px rgba(0,0,0,.3)'
handle['style']['cursor'] = 'nwse-resize'
box.appendChild(handle)

# Hover animation
box.addEventListener('mouseenter', ->(e) {
  box['style']['boxShadow'] = '0 14px 32px rgba(0,0,0,.32)'
})
box.addEventListener('mouseleave', ->(e) {
  box['style']['boxShadow'] = '0 10px 25px rgba(0,0,0,.25)'
})

# Click / Tap
box.addEventListener('click', ->(e) { console.log('clicked') })
box.addEventListener('touchstart', ->(e) { console.log('tap') })

# Drag logic
dragging = false
dragPointerId = nil
startX = 0.0
startY = 0.0
currentX = 0.0
currentY = 0.0
pendingFrame = false

window_obj = nil
begin
  window_obj = JS.global[:window]
rescue NameError
  window_obj = nil
end
raf_fn = nil
if window_obj
  raf_candidate = window_obj[:requestAnimationFrame]
  raf_fn = raf_candidate if raf_candidate
end

apply_transform = nil
apply_transform = proc do |_timestamp|
  pendingFrame = false
  box['style']['transform'] = "translate3d(#{currentX}px, #{currentY}px, 0)"
end

schedule_transform = proc do
  if raf_fn
    unless pendingFrame
      pendingFrame = true
      raf_fn.call(->(ts) { apply_transform.call(ts) })
    end
  else
    apply_transform.call(nil)
  end
end

base_shadow = box['style']['boxShadow']
drag_shadow = '0 8px 18px rgba(0,0,0,.22)'
drop_check = nil

startDrag = ->(event) {
  dragging = true
  dragPointerId = event['pointerId']
  if box['setPointerCapture']
    box['setPointerCapture'].call(dragPointerId)
  end
  box['style']['cursor'] = 'grabbing'
  box['style']['boxShadow'] = drag_shadow
  box['style']['transition'] = 'transform 0s, width .2s ease, height .2s ease, box-shadow .2s ease'
  startX = event['clientX'] - currentX
  startY = event['clientY'] - currentY
  if event['preventDefault']
    event['preventDefault'].call()
  end
}

moveDrag = ->(event) {
  if dragging && dragPointerId && event['pointerId'] == dragPointerId
    currentX = event['clientX'] - startX
    currentY = event['clientY'] - startY
    schedule_transform.call
  end
}

stopDrag = -> {
  if dragging
    dragging = false
    if dragPointerId && box['releasePointerCapture']
      box['releasePointerCapture'].call(dragPointerId)
    end
    dragPointerId = nil
    box['style']['cursor'] = 'grab'
    box['style']['boxShadow'] = base_shadow
    box['style']['transition'] = default_transition
    schedule_transform.call
  end
}

box.addEventListener('pointerdown', ->(event) {
  if !resizing
    button = event['button']
    if button == nil || button == 0
      startDrag.call(event)
    end
  end
})

# Resize logic
resizing = false
resizePointerId = nil
startW = 0
startH = 0
resizeOffsetX = 0.0
resizeOffsetY = 0.0
resizePrevTransition = ''

handle.addEventListener('pointerdown', ->(event) {
  resizing = true
  resizePointerId = event['pointerId']
  if handle['setPointerCapture']
    handle['setPointerCapture'].call(resizePointerId)
  end
  startX = event['clientX']
  startY = event['clientY']
  startW = box['offsetWidth']
  startH = box['offsetHeight']
  resizeOffsetX = currentX
  resizeOffsetY = currentY
  resizePrevTransition = box['style']['transition'] || default_transition
  box['style']['transition'] = 'none'
  box['style']['boxShadow'] = drag_shadow
  if event['stopPropagation']
    event['stopPropagation'].call()
  end
  if event['preventDefault']
    event['preventDefault'].call()
  end
})

document.addEventListener('pointermove', ->(event) {
  if resizing && resizePointerId && event['pointerId'] == resizePointerId
    dw = event['clientX'] - startX
    dh = event['clientY'] - startY
    newW = [100.0, startW.to_f + dw].max
    newH = [80.0, startH.to_f + dh].max
    currentX = resizeOffsetX + ((newW - startW.to_f) / 2.0)
    currentY = resizeOffsetY + ((newH - startH.to_f) / 2.0)
    box['style']['width'] = "#{newW.to_i}px"
    box['style']['height'] = "#{newH.to_i}px"
    schedule_transform.call
  elsif dragging && dragPointerId && event['pointerId'] == dragPointerId
    moveDrag.call(event)
  end
})

cleanup_resize = -> {
  resizing = false
  if handle['releasePointerCapture'] && resizePointerId
    handle['releasePointerCapture'].call(resizePointerId)
  end
  resizePointerId = nil
  box['style']['transition'] = resizePrevTransition || default_transition
  box['style']['boxShadow'] = base_shadow
  schedule_transform.call
}

document.addEventListener('pointerup', ->(event) {
  if resizing && resizePointerId && event['pointerId'] == resizePointerId
    cleanup_resize.call
    if event['preventDefault']
      event['preventDefault'].call()
    end
  elsif dragging && dragPointerId && event['pointerId'] == dragPointerId
    stopDrag.call
    if drop_check
      drop_check.call(event['clientX'], event['clientY'])
    end
  end
})

document.addEventListener('pointercancel', ->(event) {
  if resizing && resizePointerId && event['pointerId'] == resizePointerId
    cleanup_resize.call
  elsif dragging && dragPointerId && event['pointerId'] == dragPointerId
    stopDrag.call
  end
})

# Simple drop zone
dropzone = document.createElement('div')
dropzone['style']['position'] = 'absolute'
dropzone['style']['top'] = '20px'
dropzone['style']['left'] = '20px'
dropzone['style']['width'] = '120px'
dropzone['style']['height'] = '80px'
dropzone['style']['border'] = '2px dashed rgba(255,255,255,.75)'
dropzone['style']['borderRadius'] = '8px'
dropzone['style']['color'] = '#fff'
dropzone['style']['display'] = 'flex'
dropzone['style']['alignItems'] = 'center'
dropzone['style']['justifyContent'] = 'center'
dropzone['innerText'] = 'Drop here'
container.appendChild(dropzone)

isOver = ->(el, x, y) {
  r = el['getBoundingClientRect'].call()
  x >= r['left'] && x <= r['right'] && y >= r['top'] && y <= r['bottom']
}

drop_check = ->(clientX, clientY) {
  if isOver.call(dropzone, clientX, clientY)
    currentX = 20
    currentY = 20
    schedule_transform.call
    console.log('dropped')
  end
}

File: examples/DOM/dom_interactions.js
let box;
let boxX;
let boxY;
let container;
let dragging;
let dropzone;
let handle;
let isOver;
let label;
let moveDrag;
let parse_translate;
let resizing;
let startDrag;
let startH;
let startW;
let startX;
let startY;
let stopDrag;

const __rubyMinus = (left, right) => {
  if (Array.isArray(left)) {
    const rightValues = Array.isArray(right) ? right : [right];
    return left.filter((item) => rightValues.every((other) => other !== item));
  }
  const leftNumber = Number(left);
  const rightNumber = Number(right);
  if (!Number.isNaN(leftNumber) && !Number.isNaN(rightNumber)) {
    return leftNumber - rightNumber;
  }
  return left - right;
};

const __rubyMatch = (value, pattern) => {
  let input = value;
  let regex = pattern;
  if (value instanceof RegExp || (value && typeof value === "object" && typeof value.exec === "function")) {
    regex = value;
    input = pattern;
  }
  const normalizedInput = String(input ?? "");
  const normalizedRegex = regex instanceof RegExp ? regex : new RegExp(String(regex));
  const result = normalizedInput.match(normalizedRegex);
  if (!result) return null;
  const wrapper = {
    captures: () => result.slice(1)
  };
  result.slice(1).forEach((value, index) => { wrapper[index] = value; });
  if (result.groups && typeof result.groups === "object") {
    for (const [key, value] of Object.entries(result.groups)) {
      wrapper[key] = value;
    }
  }
  return wrapper;
};

const __rubySend = (receiver, methodName, args = [], block) => {
  if (receiver == null) return undefined;
  const fn = receiver[methodName];
  if (typeof fn === "function") {
    const callArgs = block === undefined ? args : [...args, block];
    return fn.apply(receiver, callArgs);
  }
  if (typeof methodName === "string") {
    if (methodName === "tap") {
      if (typeof block === "function") {
        const restore = typeof block.__rubyBind === "function" ? block.__rubyBind(receiver) : null;
        try { block.call(receiver, receiver); } finally { if (typeof restore === "function") restore(); }
      }
      return receiver;
    }
    if (methodName === "then" || methodName === "yield_self") {
      if (typeof block === "function") {
        const restore = typeof block.__rubyBind === "function" ? block.__rubyBind(receiver) : null;
        try { return block.call(receiver, receiver); } finally { if (typeof restore === "function") restore(); }
      }
      return receiver;
    }
    if (methodName === "catch") {
      const tag = args[0];
      if (typeof block !== "function") return undefined;
      const restore = typeof block.__rubyBind === "function" ? block.__rubyBind(receiver) : null;
      try {
        return block.call(receiver);
      } catch (error) {
        if (error && error.__rubyThrowTag !== undefined) {
          if (tag === undefined || tag === error.__rubyThrowTag || String(tag) === String(error.__rubyThrowTag)) {
            return error.__rubyThrowValue;
          }
        }
        throw error;
      } finally {
        if (typeof restore === "function") restore();
      }
    }
    if (methodName === "throw") {
      const tag = args[0];
      const value = args.length > 1 ? args[1] : undefined;
      const error = new Error("throw");
      error.__rubyThrowTag = tag;
      error.__rubyThrowValue = value;
      throw error;
    }
  }
  const missing = receiver.method_missing;
  if (typeof missing === "function") {
    const missingArgs = block === undefined ? [methodName, ...args] : [methodName, ...args, block];
    return missing.apply(receiver, missingArgs);
  }
  throw new Error(`NoMethodError: undefined method ${methodName}`);
};

const __rubyToInteger = (value) => {
  const str = String(value ?? "").trimStart();
  const match = str.match(/^[+-]?\d+/);
  if (!match) return 0;
  const parsed = parseInt(match[0], 10);
  return Number.isNaN(parsed) ? 0 : parsed;
};

const __rubyToFloat = (value) => {
  const num = parseFloat(String(value ?? ""));
  return Number.isNaN(num) ? 0 : num;
};

container = __rubySend(document, "createElement", ["div"], undefined);
container["style"]["width"] = "100%";
container["style"]["height"] = "100vh";
container["style"]["margin"] = "0";
container["style"]["display"] = "flex";
container["style"]["alignItems"] = "center";
container["style"]["justifyContent"] = "center";
container["style"]["background"] = "linear-gradient(135deg, #0ea5e9 0%, #6366f1 100%)";
container["style"]["position"] = "relative";
__rubySend(document["body"], "appendChild", [container], undefined);
box = __rubySend(document, "createElement", ["div"], undefined);
box["style"]["width"] = "220px";
box["style"]["height"] = "140px";
box["style"]["borderRadius"] = "16px";
box["style"]["boxShadow"] = "0 10px 25px rgba(0,0,0,.25)";
box["style"]["background"] = "linear-gradient(135deg, #6EE7F9 0%, #9333EA 100%)";
box["style"]["position"] = "relative";
box["style"]["cursor"] = "grab";
box["style"]["userSelect"] = "none";
box["style"]["transform"] = "translate(0px, 0px)";
box["style"]["transition"] = "transform .2s ease, width .2s ease, height .2s ease, box-shadow .2s ease";
__rubySend(container, "appendChild", [box], undefined);
label = __rubySend(document, "createElement", ["div"], undefined);
label["style"]["color"] = "#fff";
label["style"]["fontFamily"] = "system-ui, -apple-system, Segoe UI, Roboto, sans-serif";
label["style"]["fontSize"] = "14px";
label["style"]["padding"] = "12px";
label["innerText"] = "Click / Tap / Drag me";
__rubySend(box, "appendChild", [label], undefined);
handle = __rubySend(document, "createElement", ["div"], undefined);
handle["style"]["position"] = "absolute";
handle["style"]["right"] = "6px";
handle["style"]["bottom"] = "6px";
handle["style"]["width"] = "14px";
handle["style"]["height"] = "14px";
handle["style"]["borderRadius"] = "3px";
handle["style"]["background"] = "rgba(255,255,255,.85)";
handle["style"]["boxShadow"] = "0 1px 3px rgba(0,0,0,.3)";
handle["style"]["cursor"] = "nwse-resize";
__rubySend(box, "appendChild", [handle], undefined);
__rubySend(box, "addEventListener", ["mouseenter", (e) => {
  return box["style"]["boxShadow"] = "0 14px 32px rgba(0,0,0,.32)";
}], undefined);
__rubySend(box, "addEventListener", ["mouseleave", (e) => {
  return box["style"]["boxShadow"] = "0 10px 25px rgba(0,0,0,.25)";
}], undefined);
__rubySend(box, "addEventListener", ["click", (e) => {
  return __rubySend(console, "log", ["clicked"], undefined);
}], undefined);
__rubySend(box, "addEventListener", ["touchstart", (e) => {
  return __rubySend(console, "log", ["tap"], undefined);
}], undefined);
dragging = false;
startX = 0;
startY = 0;
boxX = 0;
boxY = 0;
parse_translate = (t) => {
  let m;
  m = __rubyMatch(new RegExp("translate\\(([-\\d.]+)px, ([-\\d.]+)px\\)", ""), t);
  if (m) {
    return [__rubyToFloat(m[1]), __rubyToFloat(m[2])];
  }
  else {
    return [0, 0];
  }
};
startDrag = (clientX, clientY) => {
  let coords;
  dragging = true;
  box["style"]["cursor"] = "grabbing";
  startX = clientX;
  startY = clientY;
  coords = parse_translate(box["style"]["transform"] || "");
  boxX = coords[0];
  return boxY = coords[1];
};
moveDrag = (clientX, clientY) => {
  let dx;
  let dy;
  if (!dragging) {
    return;
  }
  dx = __rubyMinus(clientX, startX);
  dy = __rubyMinus(clientY, startY);
  return box["style"]["transform"] = `translate(${__rubyToInteger(boxX + dx)}px, ${__rubyToInteger(boxY + dy)}px)`;
};
stopDrag = () => {
  dragging = false;
  return box["style"]["cursor"] = "grab";
};
__rubySend(box, "addEventListener", ["mousedown", (e) => {
  startDrag(e["clientX"], e["clientY"]);
  return e["preventDefault"]();
}], undefined);
__rubySend(document, "addEventListener", ["mousemove", (e) => {
  return moveDrag(e["clientX"], e["clientY"]);
}], undefined);
__rubySend(document, "addEventListener", ["mouseup", (e) => {
  return stopDrag();
}], undefined);
__rubySend(box, "addEventListener", ["touchstart", (e) => {
  let t;
  t = e["touches"][0];
  return startDrag(t["clientX"], t["clientY"]);
}], undefined);
__rubySend(document, "addEventListener", ["touchmove", (e) => {
  let t;
  t = e["touches"][0];
  return moveDrag(t["clientX"], t["clientY"]);
}], undefined);
__rubySend(document, "addEventListener", ["touchend", (e) => {
  return stopDrag();
}], undefined);
resizing = false;
startW = 0;
startH = 0;
__rubySend(handle, "addEventListener", ["mousedown", (e) => {
  resizing = true;
  startX = e["clientX"];
  startY = e["clientY"];
  startW = box["offsetWidth"];
  startH = box["offsetHeight"];
  e["stopPropagation"]();
  return e["preventDefault"]();
}], undefined);
__rubySend(document, "addEventListener", ["mousemove", (e) => {
  let dh;
  let dw;
  let h;
  let w;
  if (resizing) {
    dw = __rubyMinus(e["clientX"], startX);
    dh = __rubyMinus(e["clientY"], startY);
    w = __rubyToInteger(Math.max(...[100, startW + dw]));
    h = __rubyToInteger(Math.max(...[80, startH + dh]));
    box["style"]["width"] = `${w}px`;
    return box["style"]["height"] = `${h}px`;
  }
}], undefined);
__rubySend(document, "addEventListener", ["mouseup", (e) => {
  return resizing = false;
}], undefined);
dropzone = __rubySend(document, "createElement", ["div"], undefined);
dropzone["style"]["position"] = "absolute";
dropzone["style"]["top"] = "20px";
dropzone["style"]["left"] = "20px";
dropzone["style"]["width"] = "120px";
dropzone["style"]["height"] = "80px";
dropzone["style"]["border"] = "2px dashed rgba(255,255,255,.75)";
dropzone["style"]["borderRadius"] = "8px";
dropzone["style"]["color"] = "#fff";
dropzone["style"]["display"] = "flex";
dropzone["style"]["alignItems"] = "center";
dropzone["style"]["justifyContent"] = "center";
dropzone["innerText"] = "Drop here";
__rubySend(container, "appendChild", [dropzone], undefined);
isOver = (el, x, y) => {
  let r;
  r = el["getBoundingClientRect"]();
  return x >= r["left"] && x <= r["right"] && y >= r["top"] && y <= r["bottom"];
};
__rubySend(document, "addEventListener", ["mouseup", (e) => {
  if (isOver(dropzone, e["clientX"], e["clientY"])) {
    box["style"]["transform"] = "translate(20px, 20px)";
    return __rubySend(console, "log", ["dropped"], undefined);
  }
}], undefined);